### PR TITLE
refactor Optimizer class from DNN.train

### PR DESCRIPTION
**Pull request by haydnv**

### CODE REVIEW

**Comment by KOHbDS** [comment]
Maybe it would name sense to make the architecture similar to what Keras and PyTorch are doing? In particular, that would mean the following:

1. It should be the responsibility of the `Layer` to compute its gradients. Maybe all layers should have some kind of a `backward` method that computes the gradients of the weights. Then the container of the layers (probably `Model` or `NeuralNet`) should be responsible for getting the gradients of all its layers, so it should probably also have `backward` method.

2. It should be the responsibility of the optimizer to compute *and apply* the gradients. Preferably this should be done in one method call, so we need to pass a list or dict of all the parameters to be optimized to the optimizer. Some optimizers compute per-parameter statistics, so each parameter should have a unique name.

**Comment by haydnv** [comment]
I refactored Layer to compute gradients and Optimizer to compute and apply gradients in commit d04b195d. You should be able to use similar logic to `GradientDescent.optimize` to initialize tensors to track per-parameter statistics in `AdamOptimizer`.

I'm also open to replacing the `inputs` parameter with a `tape` that works like TensorFlow's `GradientTape` (see https://www.tensorflow.org/guide/autodiff), using the `hex_id` function in `tinychain.util` to assign a unique ID to each gradient to optimize within a given training step. This would be more efficient but also substantially more work to implement.

File: client/tinychain/ml/dnn.py [hunk 1]
```diff
@@ -40,24 +42,19 @@ def activation(self):
...
-        A = self.activation.forward(Z) + self["bias"]
-        return A, Z
+        weight_gradients = einsum("kj,ki->ij", [dZ, inputs])
+        bias_gradients = dZ.sum(0)
 
-    def update(self, d_weights, d_bias):
-        new_weights = self["weights"] - d_weights
-        new_bias = self["bias"] - d_bias
-        return self["weights"].write(new_weights), self["bias"].write(new_bias)
+        delta = optimizer.optimize(i, {"weights": weight_gradients, "bias": bias_gradients})
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/ml/dnn.py]: You can use the `hex_id` function in `tinychain.util` for this

File: client/tinychain/ml/dnn.py [hunk 2]
```diff
@@ -27,6 +25,10 @@ def load(cls, weights, bias, activation=Sigmoid()):
         """Load a `DNNLayer` with the given `weights` and `bias` tensors."""
 
         class _DNNLayer(cls):
+            @classmethod
+            def shape(cls):
+                return {"weights": weights.shape(), "bias": bias.shape()}
+
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/ml/dnn.py]: See above

File: client/tinychain/ml/optimizer.py [hunk 3]
```diff
@@ -27,13 +21,21 @@ def create(cls, learning_rate=F32(0.01)):
...
         return cls({"lr": learning_rate})
 
-    def optimize(self, i, gradients):
-        if isinstance(gradients, collections.abc.Mapping):
-            return {name: self.optimize(i, gradients[name]) for name in gradients}
-        elif isinstance(gradients, collections.abc.Iterable):
-            return [self.optimize(i, gradient) for gradient in gradients]
-        else:
-            return gradients * self["lr"]
+    def optimize(self, i, gradient, inputs, loss):
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/ml/optimizer.py:24]: Great suggestion! I think this could avoid double-computing the layer outputs also (currently they're computed once on the forward pass and once on the backward pass). Per offline conversaion, I'll merge in this PR and you can get started on refactoring `Differentiable` to provide a list of parameters for an `Optimizer` to optimize.

File: client/tinychain/ml/optimizer.py [hunk 4]
```diff
@@ -27,13 +21,21 @@ def create(cls, learning_rate=F32(0.01)):
...
+
+        def update(gradient, deltas):
+            shape = gradient.shape
+
+            if isinstance(shape, dict):
+                return {name: update(gradient[name], deltas[name]) for name in shape}
+            elif isinstance(shape, list) or isinstance(shape, tuple):
+                return [update(gradient[n], deltas[n]) for n in range(len(shape))]
+            else:
+                return gradient.write(deltas * lr)
```
[the remark this answers is not in the document]
Comment (haydnv) [reply client/tinychain/ml/optimizer.py]: Good catch!
